### ISSUE TITLE
IO::Buffer#write no longer works with async-2.5.0

### ISSUE THREAD

**Issue by u3shit** [socketry/async#224]
I've recently updated my ruby to 3.2.1 (from 3.1.3) and async to 2.5.0 (from 2.0.1), and code using `IO::Buffer` broke. A simple snippet like this:

```ruby
require 'async'
Async do
  b = IO::Buffer.new 16
  b.write $stdout, b.size
end
```

Throws me an error:
```
foo.rb:3: warning: IO::Buffer is experimental and both the Ruby and C interface may change in the future!
  0.0s     warn: Async::Task [oid=0x3c] [ec=0x50] [pid=9192] [2023-03-24 20:52:58 +0100]
               | Task may have ended with unhandled exception.
               |   RuntimeError: Length exceeds size of buffer!
               |   → /home/u3/.gem/ruby/3.2.0/gems/async-2.5.0/lib/async/scheduler.rb:183 in `io_write'
               |     /home/u3/.gem/ruby/3.2.0/gems/async-2.5.0/lib/async/scheduler.rb:183 in `io_write'
               |     foo.rb:4 in `write'
               |     foo.rb:4 in `block in <main>'
               |     /home/u3/.gem/ruby/3.2.0/gems/async-2.5.0/lib/async/task.rb:158 in `block in run'
               |     /home/u3/.gem/ruby/3.2.0/gems/async-2.5.0/lib/async/task.rb:310 in `block in schedule'
```

If I replace the body with
```ruby
  b = IO::Buffer.for 'abcdef'
  b.write $stdout, 2, 0
```
it prints out `bcdef`.

I've tried to run this code with `IO_EVENT_SELECTOR=EPoll` thinking that maybe the new uiring implementation is buggy, but no change. `IO_EVENT_SELECTOR=Select` ends up in an infinite loop.

**Comment by ioquatix** [comment]
Thanks, I'll take a look.

**Comment by u3shit** [comment]
Um, I think the bug is inside ruby actually.
Here it calls `rb_fiber_scheduler_io_write` with the `size_t`s converted to `VALUE`s:
https://github.com/ruby/ruby/blob/400ccb16eefe4e21c4e3eacab4fd0f208fc5e151/io_buffer.c#L2607
But that function expects `size_t`s:
https://github.com/ruby/ruby/blob/400ccb16eefe4e21c4e3eacab4fd0f208fc5e151/scheduler.c#L542
(And probably the same for the other scheduler methods.)
But at least now the last `bcdef` output makes sense, offset is 1, length is 2*2+1=5.

**Comment by ioquatix** [comment]
Ah damn, you are right, this looks like a bug in CRuby.

**Comment by ioquatix** [comment]
Sorry about this, the fix is in the linked PR. I'll try to get it backported.

**Comment by ioquatix** [comment]
It looks like this code needs more work. The linked PR fixes the reported bugs, but I'll add more tests and update the documentation.

**Comment by ioquatix** [comment]
Okay, so it was fine on Ruby 3.1 but broke it in Ruby 3.2 :(

I've merged the fix and asked for it to be backported, and added tests in <https://github.com/ruby/ruby/pull/7594>.

**Comment by u3shit** [comment]
Thanks, I've applied your two PRs on the 3.2.1 release, it seems to fix the problem.
However while reading the documentation I've noticed something.
https://ruby-doc.org/3.2.1/IO/Buffer.html#method-i-write
Here it says that `IO::Buffer#write` will write `length` bytes, and return the number of bytes written (which can be less than requested because without scheduler it just calls the unix `write` function, but that's not evident from the documentation).
https://ruby-doc.org/3.2.1/Fiber/Scheduler.html#method-i-io_write
Now, in `Scheduler#io_write` (which is called by `IO::Buffer#write` when you have a scheduler), `length` is suddenly a minimum length and the function is allowed to write out the whole buffer (and this also means the function can't return a length smaller than `length` unless an error occured. But that's also problematic, what happens for example if the calling code requires at least 16 bytes to be written out of a 32bytes buffer, the first `write` syscall can write out 10 bytes in a non-blocking fashion? Then it will block the fiber, but if after resuming the next write fails, it can't report eiher the error or the amount of data written. The current `io-event` implementation seems to prefer raising the error instead of reporting the already written out bytes.).

And indeed, running this:
```ruby
b = IO::Buffer.for 'abcdef'
b.write $stdout, 2
```
prints `ab`, but if I wrap it inside async:
```ruby
Async do
  b = IO::Buffer.for 'abcdef'
  b.write $stdout, 2
end
```
it suddenly outputs `abcdef`.

**Comment by ioquatix** [comment]
Yep this is a known problem and I will fix it.

**Comment by u3shit** [comment]
To be honest, right now my biggest problem is that if you have a code like

```ruby
buffer.write(fd, size)
```

then `size` is a minimum size when you're running under a non-blocking Fiber (like in async), while it's a maximum size otherwise. Which means the only safe way to use `IO::Buffer#write` is to slice the buffer, specify `buffer.size` as size and handle the possible short write with a loop, which is the worst of both worlds IMHO.

I have no problems with using `slice`, that's what I'd do if I were designing a C++ api with `string_view` for example. And actually my code now only even has calls like `buf.write(fd, buf.size)`, i.e. writing full buffers, I wouldn't mind if the size argument would disappear. Or at least make it optional, so you could simply write `<some_random_expression>.write(fd)` without saving it to a temporary variable (provided it guarantees writing the full buffer).

Read is a bit different beast, normal `IO#write` takes an object to write but no size argument and returns the written bytes, while `IO#read` takes a size and returns the read string (if we ignore the optional `out_string` argument). I don't think we need to force symmetry on buffer's counterparts...

About lengths and min/max/exact: in high level code and when writing, you probably only ever need to write full buffers (so the absolute one). Read is a bit trickier, as it depends on the format you read (if you have length + data on the input, you need absolute size, if you have to read until a terminating character, you want to get as much data you can in the buffer in one go and work on that, so the minimum one). In lower-level code however it might be better to be close as close as possible to the underlying unix syscalls, and handle short reads/writes however it is appropriate for you (retry immediately, so some other action before retrying, implement an event loop, whatever). There is probably no point doing that when you're using async or some other even loop library, but can be useful when you're using `IO::Buffers` on your own.

**Comment by ioquatix** [comment]
I'm reviewing the code and you are correct, it's hard to use correctly. I'll make a PR, I would appreciate your review.

**Comment by u3shit** [comment]
Sorry for the late reply. I'm not too familiar with the ruby internals, but I've tried to write something usable. Generally it looks okay for me, at least now the fiber and non-fiber operation is consistent.

**Comment by ioquatix** [comment]
Great, thanks for trying it out.

**Comment by ioquatix** [comment]
I hope this will be partly fixed in 3.2.3 and completely fixed in 3.3 (pread/pwrite, etc).
